lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1116,21 +1115,63 @@ export class UserAgentApplication {
    * @ignore
    * @hidden
    */
-  private loadFrame(urlNavigate: string, frameName: string): void {
+  private loadAndRemoveFrame(urlNavigate: string, frameName: string): void {
     // This trick overcomes iframe navigation in IE
     // IE does not load the page consistently in iframe
     this._logger.info("LoadFrame: " + frameName);
     const frameCheck = frameName;
+    const frameHandle = this.addAdalFrame(frameCheck);
+    // Check hash on iframe
+    // 100ms delay each check for 5 times. 500ms total delay before function times out.
+    frameHandle.onload = () => {
+      this.recursiveDelayCheckFrameHash(frameHandle, 5);
+    };
 
     // TODO: VSTS AI, work on either removing the 500ms timeout or making it optional for IE??
-    setTimeout(() => {
-      const frameHandle = this.addAdalFrame(frameCheck);
-      if (frameHandle.src === "" || frameHandle.src === "about:blank") {
-        frameHandle.src = urlNavigate;
-        this._logger.infoPii("Frame Name : " + frameName + " Navigated to: " + urlNavigate);
+    if (frameHandle.src === "" || frameHandle.src === "about:blank") {
+      frameHandle.src = urlNavigate;
+      this._logger.infoPii(
+        "Frame Name : " + frameName + " Navigated to: " + urlNavigate
+      );
+    }
+  }
+
+  /**
+   * Recursively check if iFrame has been redirected with hash
+   * Each check has 100ms delay
+   * @ignore
+   * @hidden
+   */
+  private recursiveDelayCheckFrameHash(
+    frameHandle: HTMLIFrameElement,
+    executionsNumber: number
+  ): void {
+    const timeout = setTimeout(() => {
+      const urlHash =
+        frameHandle.contentWindow &&
+        frameHandle.contentWindow.location &&
+        frameHandle.contentWindow.location.hash;
+      const isCallback = this.isCallback(urlHash);
+
+      // If redirect is successfully and we have the hash
+      if (isCallback) {
+        // clear timeout and proceed
+        clearTimeout(timeout);
+
+        // clear iframe so we are not loading any scripts
+        const frameDoc =
+          frameHandle.contentDocument || frameHandle.contentWindow.document;
+        frameDoc.documentElement.innerHTML = "";
```

Comment:
is frameDoc ever null

Reply:
I don't think so. iFrame should always have a document. It's a DOM element.